components/viaduct/src/backend.rs

```diff
@@ -47,6 +47,8 @@ pub fn validate_request(request: &crate::Request) -> Result<(), crate::Error> {
     if request.url.scheme() != "https"
         && request.url.host_str() != Some("localhost")
         && request.url.host_str() != Some("127.0.0.1")
+        && request.url.host_str() != Some("[::1]")
```

Comment:
In the issue, I think Martin was suggesting that the `host_str` checks can be replaced with something like:

```rust
match request.url.host() {
    Some(url::Host::Domain(d)) => d != "localhost",
    Some(url::Host::Ipv4(addr)) => !addr.is_loopback(),
    Some(url::Host::Ipv6(addr)) => !addr.is_loopback(),
    None => true,
}
```

(with a `use url` at the top of the file) which is cleaner and handles every possible way to spell an ipv4 and ipv6 loopback addresses)

Reply:
i got it but when i do this changes this validation start to fail maybe i did something wrong?
```rust
pub fn validate_request(request: &crate::Request) -> Result<(), crate::Error> {
    match request.url.host() {
        Some(url::Host::Domain(d)) => d != "localhost",
        Some(url::Host::Ipv4(addr)) => !addr.is_loopback(),
        Some(url::Host::Ipv6(addr)) => !addr.is_loopback(),
        None => return Err(crate::Error::NonTlsUrl),
    };
    Ok(())
}
// the only test that fails


        let _http_request = crate::Request::new(
            crate::Method::Get,
            url::Url::parse("http://www.example.com").unwrap(),
        );
        assert!(validate_request(&_http_request).is_err());
```

Reply:
You still need the check for `request.url.scheme() != "https"` - it's only the `host_str` checks that can be replaced with that block. That's why my block has `None => true` - ie, something like:

```rust
    if request.url.scheme() != "https"
        && match request.url.host() {
            Some(url::Host::Domain(d)) => d != "localhost",
            Some(url::Host::Ipv4(addr)) => !addr.is_loopback(),
            Some(url::Host::Ipv6(addr)) => !addr.is_loopback(),
            None => true,
        }
    {
        return Err(crate::Error::NonTlsUrl);
    }
```

Reply:
OMG, I've never tought using match this way. This really simple thing that you have shown me made me love even more this language